### ISSUE TITLE
Compaction - cardinality estimator

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#73]
When available, use the the cardinality estimator's "union" feature to estimate the number of the partitions in the compaction result, instead of the sum I used. Note that for leveled compaction, the number of partitions will be more-or-less fixed, not related to the cardinality estimator - but perhaps we should take a look at what Cassandra did here.

http://www.datastax.com/dev/blog/improving-compaction-in-cassandra-with-cardinality-estimation

**Comment by nyh** [comment]
@raphaelsc wasn't this already done, I think by you, many years ago? Or is there something missing before we can close this issue?

**Comment by raphaelsc** [comment]
> @raphaelsc wasn't this already done, I think by you, many years ago? Or is there something missing before we can close this issue?

This is less of a problem with ICS/LCS since we estimate number of partitions per sstable, so even if we cut input by half, each output fragment (but the last) will still be receiving a good estimation.

We haven't wired cardinality metadata into estimation yet, and one of the concerns raised was the algorithmic complexity to join cardinality bitmaps from all input sstables. I don't remember the exact details though.

**Comment by avikivity** [comment]
Addition vs union isn't very important. It may overestimate by a factor of nr_inputs. Let's say it's 4. We fold twice after compaction and we're done.
